File: utilslib/src/main/java/com/ashlikun/utils/ui/keyboard/SoftKeyboardSizeWatchLayout.java
package com.ashlikun.utils.ui.keyboard;

import android.content.Context;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.ViewTreeObserver;
import android.widget.RelativeLayout;

import java.util.ArrayList;
import java.util.List;

/**
 * use XhsEmotionsKeyboard(https://github.com/w446108264/XhsEmoticonsKeyboard)
 * author: sj
 */
public class SoftKeyboardSizeWatchLayout extends RelativeLayout {

    private int mOldh = -1;
    private int mNowh = -1;
    protected int mScreenHeight = 0;
    protected boolean mIsSoftKeyboardPop = false;
    public SoftKeyboardSizeWatchLayout(Context context) {
        this(context, null);
    }

    public SoftKeyboardSizeWatchLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SoftKeyboardSizeWatchLayout(Context context, AttributeSet attrs,int defStyleAttr) {
        super(context, attrs,defStyleAttr);
        getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                Rect r = new Rect();
                getRootView().getWindowVisibleDisplayFrame(r);
                if (mScreenHeight == 0) {
                    mScreenHeight = r.bottom;
                } else {
                    //防止第一次的时候获取的是假的
                    mScreenHeight = Math.max(mScreenHeight, r.bottom);
                }
                mNowh = mScreenHeight - r.bottom;
                if (mOldh != -1 && mNowh != mOldh) {
                    if (mNowh > 0) {
                        mIsSoftKeyboardPop = true;
                        if (mListenerList != null) {
                            for (OnResizeListener l : mListenerList) {
                                l.onSoftPop(mNowh);
                            }
                        }
                    } else {
                        mIsSoftKeyboardPop = false;
                        if (mListenerList != null) {
                            for (OnResizeListener l : mListenerList) {
                                l.onSoftClose();
                            }
                        }
                    }
                }
                mOldh = mNowh;
            }
        });
    }

    public boolean isSoftKeyboardPop() {
        return mIsSoftKeyboardPop;
    }


    private List<OnResizeListener> mListenerList;

    public void addOnResizeListener(OnResizeListener l) {
        if (mListenerList == null) {
            mListenerList = new ArrayList<>();
        }
        mListenerList.add(l);
    }

    public interface OnResizeListener {
        /**
         * 软键盘弹起
         */
        void onSoftPop(int height);

        /**
         * 软键盘关闭
         */
        void onSoftClose();
    }
}

File: utilslib/src/main/java/com/ashlikun/utils/ui/keyboard/SoftKeyboardSizeWatchLayout.kt
package com.ashlikun.utils.ui.keyboard

import kotlin.jvm.JvmOverloads
import android.content.Context
import android.util.AttributeSet
import android.widget.RelativeLayout
import java.util.ArrayList
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.graphics.Rect

/**
 * 软键盘弹起
 */
typealias  OnSoftPop = (height: Int) -> Unit
/**
 * 软键盘关闭
 */
typealias  OnSoftClose = () -> Unit

/**
 * use XhsEmotionsKeyboard(https://github.com/w446108264/XhsEmoticonsKeyboard)
 * author: sj
 */
open class SoftKeyboardSizeWatchLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr) {
    var oldh = -1
    var nowh = -1
    var screenHeight = 0
    var isSoftKeyboardPop = false
        protected set

    /**
     * 软键盘监听
     */
    var onSoftPops: MutableList<OnSoftPop> = mutableListOf()
    var onSoftClose: MutableList<OnSoftClose> = mutableListOf()

    init {
        viewTreeObserver.addOnGlobalLayoutListener {
            val r = Rect()
            rootView.getWindowVisibleDisplayFrame(r)
            screenHeight = if (screenHeight == 0) {
                r.bottom
            } else {
                //防止第一次的时候获取的是假的
                Math.max(screenHeight, r.bottom)
            }
            nowh = screenHeight - r.bottom
            if (oldh != -1 && nowh != oldh) {
                if (nowh > 0) {
                    isSoftKeyboardPop = true
                    onSoftPops.forEach {
                        it.invoke(nowh)
                    }
                } else {
                    isSoftKeyboardPop = false
                    onSoftClose.forEach {
                        it.invoke()
                    }
                }
            }
            oldh = nowh
        }
    }
}